src/Data/Num/Linear.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+module Data.Num.Linear
+  ( Additive(..)
+  , AddIdentity(..)
+  , Group(..)
+  , Multiplicative(..)
+  , MultIdentity(..)
+  , Semiring
+  , Ring
+  , FromInteger(..)
+  , Num(..)
+  )
+  where
+
+-- TODO: underlying semigroups/monoids
+-- TODO: flesh out laws
+import qualified Prelude
+import Data.Unrestricted.Linear
+import qualified Unsafe.Linear as Unsafe
+
+-- laws: associative, commutative
+class Additive a where
+  (+) :: a ->. a ->. a
+
+-- laws: is an identity for +
+class Additive a => AddIdentity a where
+  zero :: a
+
+-- usual abelian group laws
+class AddIdentity a => Group a where
+  {-# MINIMAL negate | (-) #-}
+  negate :: a ->. a
+  negate x = zero - x
+  (-) :: a ->. a ->. a
+  x - y = x + negate y
+
+-- laws: associative
+class Multiplicative a where
+  (*) :: a ->. a ->. a
+
+-- laws: is an identity for *
+class Multiplicative a => MultIdentity a where
+  one :: a
+
+-- laws: distributivity, annihilation
+-- | Having a linear (*) means we can't short-circuit multiplication by zero
+class (AddIdentity a, MultIdentity a) => Semiring a where
+
+-- no additional laws
+class (Group a, MultIdentity a) => Ring a where
```

Comment:
I believe you meant (up to the `AdditiveGroup` thing which I put here for consistency with the above suggestion)

```suggestion
class (AdditiveGroup a, SemingRing a) => Ring a where
```